Cython/Compiler/Parsing.py

```diff
@@ -3968,6 +3975,459 @@ def p_cpp_class_attribute(s, ctx):
+    if s.sy == "*":
+        # star pattern
+        s.next()
+        target = None
+        if s.systring != "_":  # for match-case '_' is treated as a special wildcard
+            target = p_pattern_capture_target(s)
+        else:
+            s.next()
+        pattern = MatchCaseNodes.MatchAndAssignPatternNode(
+            s.position(), target = target, is_star = True
+        )
+        return pattern
+    else:
+        pattern = p_pattern(s)
+        return pattern
+
+
+def p_pattern(s):
+    # try "as_pattern" then "or_pattern"
+    # (but practically "as_pattern" starts with "or_pattern" too)
+    patterns = []
+    pos = s.position()
+    while True:
+        patterns.append(p_closed_pattern(s))
+        if s.sy == "|":
+            s.next()
+        else:
+            break
+
+    if len(patterns) > 1:
+        pattern = MatchCaseNodes.OrPatternNode(
+            pos,
+            alternatives = patterns
+        )
+    else:
+        pattern = patterns[0]
+
+    if s.sy == 'IDENT' and s.systring == 'as':
+        s.next()
+        with tentatively_scan(s) as errors:
+            pattern.as_targets.append(p_pattern_capture_target(s))
+        if errors and s.sy == "_":
+            s.next()
+            # make this a specific error
+            return Nodes.ErrorNode(errors[0].args[0], what = errors[0].args[1])
+        elif errors:
+            with tentatively_scan(s):
+                expr = p_test(s)
+                return Nodes.ErrorNode(expr.pos, what = "Invalid pattern target")
+            s.error(errors[0])
+    return pattern
+
+
+def p_closed_pattern(s):
+    """
+    The PEG parser specifies it as
+    | literal_pattern
+    | capture_pattern
+    | wildcard_pattern
+    | value_pattern
+    | group_pattern
+    | sequence_pattern
+    | mapping_pattern
+    | class_pattern
+
+    For the sake avoiding too much backtracking, we know:
+    * starts with "{" is a sequence_pattern
+    * starts with "[" is a mapping_pattern
+    * starts with "(" is a group_pattern or sequence_pattern
+    * wildcard pattern is just identifier=='_'
+    The rest are then tried in order with backtracking
+    """
+    if s.sy == 'IDENT' and s.systring == '_':
+        pos = s.position()
+        s.next()
+        return MatchCaseNodes.MatchAndAssignPatternNode(pos)
+    elif s.sy == '{':
+        return p_mapping_pattern(s)
+    elif s.sy == '[':
+        return p_sequence_pattern(s)
+    elif s.sy == '(':
+        with tentatively_scan(s) as errors:
+            result = p_group_pattern(s)
+        if not errors:
+            return result
+        return p_sequence_pattern(s)
+
+    with tentatively_scan(s) as errors:
+        result = p_literal_pattern(s)
+    if not errors:
+        return result
+    with tentatively_scan(s) as errors:
+        result = p_capture_pattern(s)
+    if not errors:
+        return result
+    with tentatively_scan(s) as errors:
+        result = p_value_pattern(s)
+    if not errors:
+        return result
```

Comment:
This looks fairly expensive. It would be nice if we could avoid some of the repeated scans (although you probably already considered that).

Reply:
I tried to spot some of the unambiguous stuff and move it to the top of the function, so unfortunately most of the easy ways of avoiding repeated scans have already been done.

It's then a trade-off between writing it how it's specified (which I think this this close to) and trying to be clever and find better ways.

I did time the parsing of the large test_patma file and it didn't seem to bad compared to all the subsequent compiler steps so I decided not to worry too much. Don't have the numbers to hand right now though